doc/development/guide/architecture.rst

```diff
@@ -0,0 +1,238 @@
+the ``QNode`` such that the quantum gradient rules of the QNodes are registered
+to the machine learning interface via a custom gradient class or function.
+
+An interface integrates QNodes with external libraries by the following:
+
+* It wraps the QNode, returning a QNode that accepts and returns the core data
+structure of the classical machine learning library (e.g., a TF tensor, Torch
+tensor, Autograd NumPy array, etc).
+* It unwraps the input data structures to simple NumPy arrays, so that the
+quantum device can execute the user's quantum function.
+* It registers the ``QNode.jacobian()`` method as a custom gradient method, so that
+the machine learning library can 'backpropagate' across the QNode, when
+integrated into a classical computation.
+
+We refer to the :ref:`intro_interfaces` page for a more in-depth introduction
+and a list of available interfaces.
+
+Key design details
+##################
+
+The following are key design details related to how PennyLane works internally.
+
+Operators
+*********
+
+Quantum operators are incorporated by the :class:`~.Operator` class which
+contains basic information about the operator (e.g. number of parameters,
+number of wires it acts on, etc.) and further convenience methods (e.g.
+:attr:`~.Operator.matrix`, :attr:`~.Operator.eigvals`.
+
+Two important subclasses of the ``Operator`` class are:
+
+* the :class:`~.Operation` class representing quantum gates and
+* the :class:`~.Observable` representing quantum observables specified for
+  measurement.
+
+Together with ``Operator``, these classes serve as base classes for quantum
+operators.
+
+Certain operators can serve as both quantum gates and observables (e.g.
+:class:`~.PauliZ`, :class:`~.PauliX`, etc.). Such classes inherit from both
+``Operation`` and ``Observable`` classes.
+
+Quantum operators are used to build quantum functions
+which are evaluated by a ``QNode`` on a bound device. Users can define such quantum
+functions by creating regular Python functions and instantiating ``Operator``
+instances in temporal order, one per line.
+
+The following is an example of this using the :func:`~.qnode` decorator and a
+valid pre-defined device (``dev``).
+
+.. code-block:: python
+
+    @qml.qnode(dev)
+    def circuit():
+        qml.PauliX(0)
+        return qml.expval(qml.PauliZ(0))
+
+This syntax of PennyLane results in the Operator instances not being "recorded"
+by the ``QNode``. This is so because there is *no association or composition
+logic* in between the ``QNode`` that is being created and the ``Operator``
+instances: the ``QNode`` does not *contain* any instances of quantum operators.
+By executing ``circuit``, instances of the ``qml.PauliX`` and ``qml.PauliZ``
+classes are created, but they are not associated with the ``QNode`` itself.
+This creates the problem of quantum functions (``circuit``) not being able to
+keep track of which quantum operations and observables are being applied.
```

Comment:
I still find this paragraph a bit difficult to read 🤔 Perhaps it makes sense to remove it altogether, and simply go straight into describing queuing and circuit construction?

Per @co9olguy's comment, perhaps it makes sense to significantly reduce the details here (including the next subsection), and just outline what is happening in bullet point form. E.g.,

* QNode circuit construction doesn't happen until the QNode is evaluated.

* On QNode evaluation, the quantum function is executed.

  - Operators are queued to the QNode on instantiation, by having `Operator.__init__` call the `Operator.queue()` method. The operators themselves queue themselves to the surrounding :class:`~.QueuingContext`.

  - Measurement functions such as `qml.expval()` are responsible for queuing observables.